funnel/models/membership.py

```diff
@@ -0,0 +1,140 @@
+# -*- coding: utf-8 -*-
+
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.ext.hybrid import hybrid_property
+
+from baseframe import __
+from coaster.sqlalchemy import StateManager, immutable, with_roles
+from coaster.utils import LabeledEnum
+
+from . import BaseMixin, UuidMixin, db
+from .user import User
+
+__all__ = ['MEMBERSHIP_RECORD_TYPE']
+
+
+class MEMBERSHIP_RECORD_TYPE(LabeledEnum):  # NOQA: N801
+    INVITE = (0, 'invite', __("Invite"))
+    ACCEPT = (1, 'accept', __("Accept"))
+    DIRECT_ADD = (2, 'direct_add', __("Direct add"))
+    AMEND = (3, 'amend', __("Amend"))
+
+
+class ImmutableMembershipMixin(UuidMixin, BaseMixin):
+    """
+    Support class for immutable memberships
+    """
+
+    __uuid_primary_key__ = True
+    #: List of columns that will be copied into a new row when a membership is amended
+    __data_columns__ = ()
+    #: Parent column ('profile_id' or 'project_id' in the subclasses)
+    __parent_column__ = None
```

Comment:
IMO, this should be replaced with a `parent_id` synonym (defined in subclasses), following the pattern established by the `BaseScoped*` mixin classes (but using `parent_id` instead of `parent`, to allow membership amendment without loading the parent object from the db).